# Final Project

## Dataset : Music & Mental Health Survey Results

------------------------------------------------------------------------------------------------------
The dataset was extracted in kaggle, and the data collection was managed by Catherine Rasgaitis, through a form with generic questions regarding musical background and listening habits and mental health symptoms.

Columns:
- Timestamp - Date and time when form was submitted
- Age - Respondent's age 
- Primary streaming service - Respondent's primary streaming service
- Hours per day - Number of hours the respondent listens to music per day
- While working - Does the respondent listen to music while 
- Instrumentalist - Does the respondent play an instrument regularly?
- Composer - Does the respondent compose music?
- Fav genre - Respondent's favorite or top genre
- Exploratory - Does the respondent actively explore new artists/genres?
- Foreign languages - Does the respondent regularly listen to music with lyrics in a language they are not fluent in?
- BPM - Beats per minute of favorite genre
- Frequency [Classical] - How frequently the respondent listens to classical music
- Frequency [Country] - How frequently the respondent listens to country music
- Frequency [EDM] - How frequently the respondent listens to EDM music
- Frequency [Folk] - How frequently the respondent listens to folks music
- Frequency [Gospel] - How frequently the respondent listens to Gospel music
- Frequency [Hip hop] - How frequently the respondent listens to hip hop music
- Frequency [Jazz] - How frequently the respondent listens to jazz music
- Frequency [K pop] - How frequently the respondent listens to K pop music
- Frequency [Latin] - How frequently the respondent listens to Latin music
- Frequency [Lofi] - How frequently the respondent listens to lofi music
- Frequency [Metal] - How frequently the respondent listens to metal music
- Frequency [Pop] - How frequently the respondent listens to pop music
- Frequency [R&B] - How frequently the respondent listens to R&B music
- Frequency [Rap] - How frequently the respondent listens to rap music
- Frequency [Rock] - How frequently the respondent listens to rock music
- Frequency [Video game music] - How frequently the respondent listens to video game music
- Anxiety - Self-reported anxiety, on a scale of 0-10
- Depression - Self-reported depression, on a scale of 0-10
- Insomnia - Self-reported insomnia, on a scale of 0-10
- OCD - Self-reported OCD, on a scale of 0-10
- Music effects - Does music improve/worsen respondent's mental health conditions?
- Permissions - Permissions to publicize data

kaggle link: https://www.kaggle.com/datasets/catherinerasgaitis/mxmh-survey-results

-----------------------------------------------------------------------------------------------------

## Objective

--------------------------------------------------------------------------------------------------------
The goal here is to identify correlations between a person's music taste or relationship with music and its mental health. And additionaly to predict if an individuals music preference can influence their mental health status.

-------------------------------------------------------------------------------------------------------

## Confirmatory Data Analysis (CDA)

In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np

In [2]:
# Loading the cleaned 'Music & Mental Health Survey Results' dataset
df = pd.read_csv('archive\survey_results_cleaned.csv', encoding="latin1", sep= ",")
df.head()

age primary_streaming_service  hours_per_day while_working instrumentalist  \
0   18                   Spotify            3.0           Yes             Yes   
1   63                   Pandora            1.5           Yes              No   
2   18                   Spotify            4.0            No              No   
3   61             YouTube Music            2.5           Yes              No   
4   18                   Spotify            4.0           Yes              No   

  composer         fav_genre exploratory foreign_languages    bpm  ...  \
0      Yes             Latin         Yes               Yes  156.0  ...   
1       No              Rock         Yes                No  119.0  ...   
2       No  Video game music          No               Yes  132.0  ...   
3      Yes              Jazz         Yes               Yes   84.0  ...   
4       No               R&B         Yes                No  107.0  ...   

          freq_pop         freq_r&b         freq_rap        freq_rock  \
0  Very frequently        Sometimes  Very frequently            Never   
1        Sometimes        Sometimes           Rarely  Very frequently   
2           Rarely            Never           Rarely           Rarely   
3        Sometimes        Sometimes            Never            Never   
4        Sometimes  Very frequently  Very frequently            Never   

  freq_video_game_music anxiety depression insomnia  ocd music_effects  
0             Sometimes     3.0        0.0      1.0  0.0     No effect  
1                Rarely     7.0        2.0      2.0  1.0     No effect  
2       Very frequently     7.0        7.0     10.0  2.0     No effect  
3                 Never     9.0        7.0      3.0  3.0       Improve  
4                Rarely     7.0        2.0      5.0  9.0       Improve  

[5 rows x 31 columns]

### 1. Hypothesis Testing

---------------------------------------------------------------------------------------------------------------------------------------------------
Let's perform a Analysis of Variance (ANOVA) to compare the means of the 4 mental health issues in question across various different music genres (using column fav_genre).

---------------------------------------------------------------------------------------------------------------------------------------------------

###### Depression

In [3]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Assuming a significant ANOVA result, conducting post hoc analysis
post_hoc = pairwise_tukeyhsd(endog=df['depression'], groups=df['fav_genre'], alpha=0.05)

print(post_hoc)

       Multiple Comparison of Means - Tukey HSD, FWER=0.05       
  group1       group2      meandiff p-adj   lower   upper  reject
-----------------------------------------------------------------
Classical          Country   0.2445    1.0 -2.2593  2.7484  False
Classical              EDM   1.1678 0.9102  -1.043  3.3786  False
Classical             Folk   0.9912 0.9877 -1.3666   3.349  False
Classical           Gospel  -1.4088 0.9994 -5.8539  3.0363  False
Classical          Hip hop   1.8069 0.3073 -0.4606  4.0744  False
Classical             Jazz   0.4245    1.0 -2.2836  3.1327  False
Classical            K pop   0.3476    1.0 -2.1233  2.8185  False
Classical            Latin  -1.0755    1.0 -7.1999  5.0489  False
Classical             Lofi   2.5245 0.5181 -1.0334  6.0825  False
Classical            Metal   0.9927 0.8745 -0.8016   2.787  False
Classical              Pop   0.4114    1.0 -1.3043   2.127  False
Classical              R&B  -0.2469    1.0 -2.4946  2.0008  False
Classical 

-------------------------------------------------------------------------------------------------------------------------
For all music genre combinatios, the null cannot be rejected. Meaning, for example:
- Given a p-adj value of 1.0 and a reject value of False, we would conclude that there's no statistically significant difference in the mean value of the measured variable (depression) between individuals who prefer Classical music and those who prefer Country music based on this test. The wide confidence interval that includes both positive and negative values further underscores the uncertainty around this meandiff estimate, suggesting that any observed difference could be due to chance.

- In summary, this particular comparison does not provide evidence to support a difference in means between the Classical and Country groups for the variable in question under the conditions of this study.

------------------------------------------------------------------------------------------------------------------------

###### Anxiety

In [4]:
# Assuming a significant ANOVA result, conducting post hoc analysis
post_hoc = pairwise_tukeyhsd(endog=df['anxiety'], groups=df['fav_genre'], alpha=0.05)

print(post_hoc)

      Multiple Comparison of Means - Tukey HSD, FWER=0.05       
  group1       group2      meandiff p-adj   lower  upper  reject
----------------------------------------------------------------
Classical          Country   0.5132    1.0 -1.8109 2.8373  False
Classical              EDM   0.5997 0.9998 -1.4524 2.6518  False
Classical             Folk   1.6799 0.3729 -0.5087 3.8684  False
Classical           Gospel  -0.0535    1.0 -4.1794 4.0725  False
Classical          Hip hop   1.2309 0.8188 -0.8739 3.3356  False
Classical             Jazz   1.0132 0.9919 -1.5005  3.527  False
Classical            K pop    1.344 0.8165 -0.9495 3.6375  False
Classical            Latin  -0.5535    1.0 -6.2382 5.1312  False
Classical             Lofi   1.2132  0.997 -2.0893 4.5157  False
Classical            Metal   0.8746  0.914 -0.7909 2.5401  False
Classical              Pop   1.1878 0.4255 -0.4047 2.7803  False
Classical              R&B   0.2846    1.0 -1.8017  2.371  False
Classical              Ra

-------------------------------------------------------------------------------------------------------------------
Again, we cannot reject the null hypothesis for any of the above combinations.

Thereby, there is no evidence to support that a difference of means of anxiety scores for each two different music genre combination.

--------------------------------------------------------------------------------------------------------------------
    

###### Insomnia

In [5]:
# Assuming a significant ANOVA result, conducting post hoc analysis
post_hoc = pairwise_tukeyhsd(endog=df['insomnia'], groups=df['fav_genre'], alpha=0.05)

print(post_hoc)

      Multiple Comparison of Means - Tukey HSD, FWER=0.05       
  group1       group2      meandiff p-adj   lower  upper  reject
----------------------------------------------------------------
Classical          Country  -1.0725 0.9881 -3.6326 1.4877  False
Classical              EDM   0.1805    1.0   -2.08  2.441  False
Classical             Folk  -0.1591    1.0 -2.5699 2.2517  False
Classical           Gospel   1.5409 0.9988 -3.0041 6.0859  False
Classical          Hip hop  -0.3219    1.0 -2.6403 1.9966  False
Classical             Jazz   0.0575    1.0 -2.7115 2.8266  False
Classical            K pop  -0.3309    1.0 -2.8574 2.1955  False
Classical            Latin  -0.4591    1.0 -6.7211 5.8029  False
Classical             Lofi   1.8075 0.9444 -1.8304 5.4455  False
Classical            Metal   0.7644 0.9887 -1.0703  2.599  False
Classical              Pop   -0.424    1.0 -2.1783 1.3302  False
Classical              R&B  -0.9067 0.9935 -3.2049 1.3915  False
Classical              Ra

-------------------------------------------------------------------------------------------------------------------
Again, we cannot reject the null hypothesis for any of the above combinations.

Thereby, there is no evidence to support that a difference of means of insomnia scores for each two different music genre combination.

--------------------------------------------------------------------------------------------------------------------

###### OCD

In [6]:
# Assuming a significant ANOVA result, conducting post hoc analysis
post_hoc = pairwise_tukeyhsd(endog=df['ocd'], groups=df['fav_genre'], alpha=0.05)

print(post_hoc)

      Multiple Comparison of Means - Tukey HSD, FWER=0.05       
  group1       group2      meandiff p-adj   lower  upper  reject
----------------------------------------------------------------
Classical          Country   0.3826    1.0 -1.9952 2.7604  False
Classical              EDM   0.6226 0.9997 -1.4769 2.7222  False
Classical             Folk  -0.1774    1.0 -2.4165 2.0618  False
Classical           Gospel   -2.044 0.9551 -6.2653 2.1773  False
Classical          Hip hop   0.2109    1.0 -1.9425 2.3642  False
Classical             Jazz   0.4226    1.0 -2.1492 2.9945  False
Classical            K pop   0.1611    1.0 -2.1854 2.5076  False
Classical            Latin  -0.7107    1.0 -6.5268 5.1054  False
Classical             Lofi   1.0226 0.9997 -2.3562 4.4015  False
Classical            Metal   0.0204    1.0 -1.6836 1.7244  False
Classical              Pop   0.4779 0.9998 -1.1514 2.1072  False
Classical              R&B   0.3655    1.0  -1.769    2.5  False
Classical              Ra

-------------------------------------------------------------------------------------------------------------------
Again, we cannot reject the null hypothesis for any of the above combinations.

Thereby, there is no evidence to support that a difference of means of OCD scores for each two different music genre combination.

--------------------------------------------------------------------------------------------------------------------

### 2. Regression Analysis

##### 2.1 Linear Regression

---------------------------------------------------------------------------------------------------------------------------------------------------------------------
We'll be using Linear Regression to predict a continuous mental health score (depression, anxiety, insomnia and OCD) from various music preferences or genres (fav_genre)

----------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [7]:
# One-hot encode 'fav_genre'
df_encoded = pd.get_dummies(df, columns=['fav_genre'], drop_first=True)

df_encoded.head()

age primary_streaming_service  hours_per_day while_working instrumentalist  \
0   18                   Spotify            3.0           Yes             Yes   
1   63                   Pandora            1.5           Yes              No   
2   18                   Spotify            4.0            No              No   
3   61             YouTube Music            2.5           Yes              No   
4   18                   Spotify            4.0           Yes              No   

  composer exploratory foreign_languages    bpm freq_classical  ...  \
0      Yes         Yes               Yes  156.0         Rarely  ...   
1       No         Yes                No  119.0      Sometimes  ...   
2       No          No               Yes  132.0          Never  ...   
3      Yes         Yes               Yes   84.0      Sometimes  ...   
4       No         Yes                No  107.0          Never  ...   

  fav_genre_Jazz fav_genre_K pop fav_genre_Latin fav_genre_Lofi  \
0          False           False            True          False   
1          False           False           False          False   
2          False           False           False          False   
3           True           False           False          False   
4          False           False           False          False   

  fav_genre_Metal fav_genre_Pop fav_genre_R&B fav_genre_Rap fav_genre_Rock  \
0           False         False         False         False          False   
1           False         False         False         False           True   
2           False         False         False         False          False   
3           False         False         False         False          False   
4           False         False          True         False          False   

  fav_genre_Video game music  
0                      False  
1                      False  
2                       True  
3                      False  
4                      False  

[5 rows x 45 columns]

In [8]:
# Step 1: Drop specific target columns to create a temporary DataFrame that still holds potential categorical columns
temp_features = df_encoded.drop(['depression', 'anxiety', 'insomnia', 'ocd'], axis=1)

# Step 2: Select only numerical columns from temp_features. This keeps one-hot encoded and numerical while dropping original categoricals.
X = temp_features.select_dtypes(include=['number'])


###### Depression

In [9]:
# Extracting the target variable remains the same
y = df_encoded['depression']

In [10]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the Linear Regression Model
model = LinearRegression()
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluation
mse = mean_squared_error(y_test, y_pred)
r_squared = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r_squared}")

Mean Squared Error: 8.178151044602116
R-squared: 0.021786164854450507


---------------------------------------------------------------------------------------------------------------------------------------
- A MSE of 8.178 indicates that, on average, the square of the difference between observed and predicted values in our model is 8.178. The smaller the MSE, the closer your model's predictions are to the actual values, indicating a better fit.

-  An R² value of approximately 0.022 suggests that about 2.2% of the variance in our dependent variable (depression) can be explained by our model. Given that R² ranges from 0 to 1, where 1 indicates a perfect fit, a value as low as 0.022 implies that our model does not explain much of the variance in the target variable, suggesting that the model has a weak predictive capability in its current form.

----------------------------------------------------------------------------------------------------------------------------------------

###### Anxiety

In [11]:
# Extracting the target variable remains the same
y = df_encoded['anxiety']

In [12]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the Linear Regression Model
model = LinearRegression()
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluation
mse = mean_squared_error(y_test, y_pred)
r_squared = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r_squared}")

Mean Squared Error: 7.939347897082469
R-squared: -0.06517183307342767


-------------------------------------------------------------------------------------------
Again we didn't get good results, which means that our model has a weak predictive capability for anxiety scores too.

-------------------------------------------------------------------------------------------

###### Insomnia

In [13]:
# Extracting the target variable remains the same
y = df_encoded['insomnia']

In [14]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the Linear Regression Model
model = LinearRegression()
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluation
mse = mean_squared_error(y_test, y_pred)
r_squared = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r_squared}")

Mean Squared Error: 8.559411075668894
R-squared: -0.03982433842654376


-------------------------------------------------------------------------------------------
Again we didn't get good results, which means that our model has a weak predictive capability for insomnia scores too.

-------------------------------------------------------------------------------------------

###### OCD

In [15]:
# Extracting the target variable remains the same
y = df_encoded['ocd']

In [16]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the Linear Regression Model
model = LinearRegression()
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Evaluation
mse = mean_squared_error(y_test, y_pred)
r_squared = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R-squared: {r_squared}")

Mean Squared Error: 9.812996882211493
R-squared: -0.005755907462017484


-------------------------------------------------------------------------------------------
Again we didn't get good results, which means that our model has a weak predictive capability for OCD scores too.

-------------------------------------------------------------------------------------------

### 2.2 Logistic Regression

----------------------------------------------------------------------------------------------------------------------------------------------------------------------
We'll be performing Logistic Regression to  reveal how changes in the frequency of listening to specific music genres (encoded numerically) influence the odds of having high levels of depression or anxiety, for example.

----------------------------------------------------------------------------------------------------------------------------------------------------------

In [18]:
# Loading the encoded frequency columns from 'Music & Mental Health Survey Results' dataset
df_freq_encoded = pd.read_csv(r'C:\Users\mafal\Documents\ironhack\projects\project-music-and-mental-health-survey\archive\freq_col_encoded.csv', encoding="latin1", sep= ",")
df_freq_encoded.head()

freq_classical_encoded  freq_country_encoded  freq_edm_encoded  \
0                       3                     1                 3   
1                       2                     1                 1   
2                       1                     1                 4   
3                       2                     1                 1   
4                       1                     1                 3   

   freq_folk_encoded  freq_gospel_encoded  freq_hip_hop_encoded  \
0                  1                    1                     2   
1                  3                    2                     3   
2                  1                    1                     3   
3                  3                    2                     1   
4                  1                    3                     4   

   freq_jazz_encoded  freq_k_pop_encoded  freq_latin_encoded  \
0                  1                   4                   4   
1                  4                   3                   2   
2                  3                   4                   1   
3                  4                   2                   4   
4                  1                   4                   2   

   freq_lofi_encoded  freq_metal_encoded  freq_pop_encoded  freq_r&b_encoded  \
0                  3                   1                 4                 2   
1                  3                   1                 2                 2   
2                  2                   2                 3                 1   
3                  2                   1                 2                 2   
4                  2                   1                 2                 4   

   freq_rap_encoded  freq_rock_encoded  freq_video_game_music_encoded  \
0                 4                  1                              2   
1                 3                  4                              3   
2                 3                  3                              4   
3                 1                  1                              1   
4                 4                  1                              3   

   anxiety  depression  insomnia  ocd  
0      3.0         0.0       1.0  0.0  
1      7.0         2.0       2.0  1.0  
2      7.0         7.0      10.0  2.0  
3      9.0         7.0       3.0  3.0  
4      7.0         2.0       5.0  9.0

###### Depression

In [19]:
# Example of binning the 'depression' scores into low (0-3), medium (4-6), and high (7-10) 
df_freq_encoded['depression_category'] = pd.cut(df_freq_encoded['depression'], bins=[-np.inf, 3, 6, np.inf], labels=['Low', 'Medium', 'High'])

# Use get_dummies to encode these categories if applying Logistic Regression
df_freq_encoded = pd.get_dummies(df_freq_encoded, columns=['depression_category'], drop_first=True)

In [21]:
df_freq_encoded

freq_classical_encoded  freq_country_encoded  freq_edm_encoded  \
0                         3                     1                 3   
1                         2                     1                 1   
2                         1                     1                 4   
3                         2                     1                 1   
4                         1                     1                 3   
..                      ...                   ...               ...   
730                       4                     3                 1   
731                       3                     3                 1   
732                       3                     2                 2   
733                       4                     1                 1   
734                       2                     3                 4   

     freq_folk_encoded  freq_gospel_encoded  freq_hip_hop_encoded  \
0                    1                    1                     2   
1                    3                    2                     3   
2                    1                    1                     3   
3                    3                    2                     1   
4                    1                    3                     4   
..                 ...                  ...                   ...   
730                  2                    1                     2   
731                  1                    1                     1   
732                  3                    3                     4   
733                  1                    1                     1   
734                  2                    3                     4   

     freq_jazz_encoded  freq_k_pop_encoded  freq_latin_encoded  \
0                    1                   4                   4   
1                    4                   3                   2   
2                    3                   4                   1   
3                    4                   2                   4   
4                    1                   4                   2   
..                 ...                 ...                 ...   
730                  3                   1                   2   
731                  3                   1                   1   
732                  3                   3                   3   
733                  3                   1                   1   
734                  4                   2                   1   

     freq_lofi_encoded  ...  freq_r&b_encoded  freq_rap_encoded  \
0                    3  ...                 2                 4   
1                    3  ...                 2                 3   
2                    2  ...                 1                 3   
3                    2  ...                 2                 1   
4                    2  ...                 4                 4   
..                 ...  ...               ...               ...   
730                  3  ...                 1                 3   
731                  3  ...                 1                 1   
732                  2  ...                 2                 2   
733                  1  ...                 1                 1   
734                  3  ...                 4                 4   

     freq_rock_encoded  freq_video_game_music_encoded  anxiety  depression  \
0                    1                              2      3.0         0.0   
1                    4                              3      7.0         2.0   
2                    3                              4      7.0         7.0   
3                    1                              1      9.0         7.0   
4                    1                              3      7.0         2.0   
..                 ...                            ...      ...         ...   
730                  4                              1      7.0         6.0   
731                  2                              2      3.0         2.0   
732                  3 

In [25]:
# X contains the predictors, which include the freq columns
# Assuming all freq columns are prefixed with 'freq_', and are already suitable for analysis
X = df_freq_encoded[[col for col in df_freq_encoded.columns if col.startswith('freq_')]]

# y contains the binary encoded depression outcome
y = df_freq_encoded['depression_category_Medium']  # Assuming this is your binary encoded depression column

In [26]:
from sklearn.model_selection import train_test_split

# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [28]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.6938775510204082
Confusion Matrix:
 [[102   1]
 [ 44   0]]


----------------------------------------------------------------------------------------
- An accuracy of approximately 0.694 means that around 69.4% of your model's predictions are correct. While this might seem decent at a glance, accuracy alone can sometimes be misleading, especially if the class distribution is imbalanced. Relying solely on accuracy might not capture the model's performance nuances.

- About the Confusion Matrix:
    - True Negatives (TN): Thetop-leftsquare (102) represents the number of correctly predicted negative class labels (e.g., not depressed).
    - False Positives (FP): The top-right square (1) shows the number of negative class labels incorrectly predicted as positive (e.g., predicted as depressed when they are not).
    - False Negatives (FN): The bottom-left square (44) shows the number of positive class labels incorrectly predicted as negative (e.g., predicted as not depressed when they are).
    - True Positives (TP): The bottom-rightsquare (0) represents the number of correctly predicted positive class labels (e.g., depressed).

Further Insights from the Confusion Matrix:
- High FN with Zero TP: Your model didn't correctly identify any of the positive cases ('depressed'), as indicated by the 0 in the TP section of the matrix. All positive cases were misclassified as negatives, which is concerning for predictive tasks aiming to identify positive cases.

- Imbalanced Data Suspected: The model seems biased toward predicting the majority class (assuming non-depressed cases are more prevalent, given the high TN), a common issue in imbalanced datasets.

------------------------------------------------------------------------------------------------------------------------

In [29]:
# y contains the binary encoded depression outcome
y = df_freq_encoded['depression_category_High'] 

In [30]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [31]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [32]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.6190476190476191
Confusion Matrix:
 [[85 11]
 [45  6]]


-------------------------------------------------------------------------
- True Negatives (TN): 85 - The count of negative instances (e.g., Non-Depressed) correctly predicted by the model.
- False Positives (FP): 11 - The count of negative instances incorrectly predicted as positive (e.g., Non-Depressed incorrectly identified as Depressed).
- False Negatives (FN): 45 - The count of positive instances (e.g., Depressed) incorrectly predicted as negative by the model.
- True Positives (TP): 6 - The count of positive instances correctly predicted by the model.

Evaluation:

- Imbalance Impact: The significantly greater number of False Negatives (45) compared to True Positives (6) suggests that the model struggles with identifying positive instances, which may indicate the model is biased towards predicting the negative class, possibly due to class imbalance.

- Accuracy Limitation: Despite the accuracy being over 61%, the breakdown in the confusion matrix highlights the model's limitations, especially in detecting the positive class effectively.

------------------------------------------------------------------------

###### Anxiety

In [33]:
# Example of binning the 'anxiety' scores into low (0-3), medium (4-6), and high (7-10) 
df_freq_encoded['anxiety_category'] = pd.cut(df_freq_encoded['anxiety'], bins=[-np.inf, 3, 6, np.inf], labels=['Low', 'Medium', 'High'])

# Use get_dummies to encode these categories if applying Logistic Regression
df_freq_encoded = pd.get_dummies(df_freq_encoded, columns=['anxiety_category'], drop_first=True)

In [34]:
# X contains the predictors, which include the freq columns
# Assuming all freq columns are prefixed with 'freq_', and are already suitable for analysis
X = df_freq_encoded[[col for col in df_freq_encoded.columns if col.startswith('freq_')]]

# y contains the binary encoded anxiety outcome
y = df_freq_encoded['anxiety_category_Medium']  # Assuming this is your binary encoded depression column

In [35]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [36]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [37]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.6938775510204082
Confusion Matrix:
 [[102   1]
 [ 44   0]]


---------------------------------------------------------------------
Again the model's failure to correctly identify any positive cases is particularly concerning if the positive class is critical to the task (e.g., identifying individuals with anxiety).

-------------------------------------------------------------------------------------

In [43]:
# y contains the binary encoded anxiety outcome
y = df_freq_encoded['anxiety_category_High']  # Assuming this is your binary encoded depression column

In [44]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [45]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [46]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.5986394557823129
Confusion Matrix:
 [[47 22]
 [37 41]]


--------------------------------------------
The confusion matrix is arranged as:

- Top-Left (True Negatives, TN): 47 - The model correctly predicted the negative class this many times.
- Top-Right (False Positives, FP): 22 - The model incorrectly predicted these cases as positive when they were actually negative.
- Bottom-Left (False Negatives, FN): 37 - The model incorrectly predicted these cases as negative when they were actually positive.
- Bottom-Right (True Positives, TP): 41 - The model correctly predicted the positive class this many times.

---------------------------------------------------------------------------

###### Insomnia

In [38]:
# Example of binning the 'insomnia' scores into low (0-3), medium (4-6), and high (7-10) 
df_freq_encoded['insomnia_category'] = pd.cut(df_freq_encoded['insomnia'], bins=[-np.inf, 3, 6, np.inf], labels=['Low', 'Medium', 'High'])

# Use get_dummies to encode these categories if applying Logistic Regression
df_freq_encoded = pd.get_dummies(df_freq_encoded, columns=['insomnia_category'], drop_first=True)

In [39]:
# X contains the predictors, which include the freq columns
# Assuming all freq columns are prefixed with 'freq_', and are already suitable for analysis
X = df_freq_encoded[[col for col in df_freq_encoded.columns if col.startswith('freq_')]]

# y contains the binary encoded insomnia outcome
y = df_freq_encoded['insomnia_category_Medium']  # Assuming this is your binary encoded depression column

In [40]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [41]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [42]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.7142857142857143
Confusion Matrix:
 [[105   0]
 [ 42   0]]


In [47]:
# y contains the binary encoded insomnia outcome
y = df_freq_encoded['insomnia_category_High']  # Assuming this is your binary encoded depression column

In [48]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [49]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [50]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.8027210884353742
Confusion Matrix:
 [[118   1]
 [ 28   0]]


---------------------------------------------------------------------
Again the model's failure to correctly identify any positive cases is particularly concerning if the positive class is critical to the task (e.g., identifying individuals with insomnia).

-------------------------------------------------------------------------------------

###### OCD

In [51]:
# Example of binning the 'ocd' scores into low (0-3), medium (4-6), and high (7-10) 
df_freq_encoded['ocd_category'] = pd.cut(df_freq_encoded['ocd'], bins=[-np.inf, 3, 6, np.inf], labels=['Low', 'Medium', 'High'])

# Use get_dummies to encode these categories if applying Logistic Regression
df_freq_encoded = pd.get_dummies(df_freq_encoded, columns=['ocd_category'], drop_first=True)

In [55]:
# X contains the predictors, which include the freq columns
# Assuming all freq columns are prefixed with 'freq_', and are already suitable for analysis
X = df_freq_encoded[[col for col in df_freq_encoded.columns if col.startswith('freq_')]]

# y contains the binary encoded ocd outcome
y = df_freq_encoded['ocd_category_Medium']  # Assuming this is your binary encoded depression column

In [56]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [57]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [58]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.8231292517006803
Confusion Matrix:
 [[121   0]
 [ 26   0]]


In [59]:
# y contains the binary encoded ocd outcome
y = df_freq_encoded['ocd_category_High']  # Assuming this is your binary encoded depression column

In [60]:
# Split the dataset into training (80%) and testing (20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [61]:
# Initialize the logistic regression model
model = LogisticRegression(max_iter=1000)  # Adjust max_iter as necessary for convergence

# Fit the model
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [62]:
# Basic evaluation using accuracy and a confusion matrix
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.8163265306122449
Confusion Matrix:
 [[120   0]
 [ 27   0]]


---------------------------------------------------------------------
Again the model's failure to correctly identify any positive cases is particularly concerning if the positive class is critical to the task (e.g., identifying individuals with ocd).

-------------------------------------------------------------------------------------

### 3. Machine Learning